### PR TITLE
Adding DUC semantic segmentation model

### PR DESCRIPTION
**Pull request by abhinavs95**
Adding DUC semantic segmentation model with the following artifacts:

* Model readme page with links to ONNX model, model with sample test data, MD5 checksum
* Notebook for inference with details of inference in model readme
* Notebook for validation with details of evaluation metric and accuracy in model readme
* Scripts for pre and post-processing
* Details of dataset and instructions on how to prepare it
* Additional scripts required by the notebooks
* Model description and use cases, references, contributors, license info in model readme
* Added link from main readme page, reorganized sections, fixed links

I am working on the training notebook and will add it soon, either in this PR or in a separate one.

Also it seems when merging the model zoos (#75 ) the commits were not transferred correctly which is why all the old commits are showing here. It should not be a problem as the old changes are already merged and the Files changed reflect only the new changes.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,12 +45,22 @@ These models detect and/or recognize human faces in images. Some more popular mo
...
+
+<hr>
+
 ### Object Detection & Segmentation
 These models detect the presence of multiple objects in an image and segment out areas of the image where the objects are detected.
 
 |Model Class |Reference |Description |
 |-|-|-|
-|<b>Tiny_YOLOv2</b>|[Redmon et al.](https://arxiv.org/pdf/1612.08242.pdf)|Deep CNN model for Object Detection|
+|<b>[Tiny_YOLOv2](tiny_yolov2)</b>|[Redmon et al.](https://arxiv.org/pdf/1612.08242.pdf)|Deep CNN model for Object Detection|
```
Comment (ankkhedia) [inline README.md]: Good catch!

File: models/semantic_segmentation/DUC/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+# DUC
+
+## Use cases
+DUC is a semantic segmentation model, i.e., for an input image the model labels each pixel in the image into a set of pre-defined categories. The model provides very good accuracy in terms of mIOU score and can be used in any application requiring semantic segmentation. In particular, since the model is trained on the [cityscapes dataset](#dset) which contains images from urban street scenes, it can be used effectively in self driving vehicle systems.
```
Comment (ankkhedia) [inline models/semantic_segmentation/DUC/README.md]: Add a section or note describing what is mIOU score

File: models/semantic_segmentation/DUC/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+### Setup
+* Unpack the zip files into folders `leftImg8bit_trainvaltest` and `gtFine_trainvaltest`.
+* Use the path to the train/val folders inside these folders for training/validation.
+* Use the files [val.lst](https://s3.amazonaws.com/onnx-model-zoo/duc/val.lst) and [train.lst](https://s3.amazonaws.com/onnx-model-zoo/duc/train.lst) containing image lists required during validation and training respectively.
+
+Please note that the dataset is under copyright. Refer to the [citation](https://www.cityscapes-dataset.com/citation/) page for details.
+
+## Validation accuracy
+The accuracies obtained by the models on the validation set are mentioned above and they match with those mentioned in the paper.
```
Comment (ankkhedia) [inline models/semantic_segmentation/DUC/README.md]: Mention that the metric definition is mentioned below in the validation section

File: models/semantic_segmentation/DUC/cityscapes_loader.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
+import atexit
+import logging
+import multiprocessing as mp
+import random
```
Comment (ankkhedia) [inline models/semantic_segmentation/DUC/cityscapes_loader.py]: Add comments saying what these functions are doing or an overall comment for file

**Comment by abhinavs95** [comment]
Thanks for the review @ankkhedia I'll push the changes soon.

@aaronmarkham @ThomasDelteil @lupesko @houseroad @mx-iao @prasanthpul could you please review the changes? Thanks!

File: models/semantic_segmentation/DUC/cityscapes_loader.py (hunk 4 above)
Comment (abhinavs95) [reply models/semantic_segmentation/DUC/cityscapes_loader.py]: done

File: models/semantic_segmentation/DUC/README.md (hunk 2 above)
Comment (abhinavs95) [reply models/semantic_segmentation/DUC/README.md]: linked to section below

File: models/semantic_segmentation/DUC/README.md (hunk 3 above)
Comment (abhinavs95) [reply models/semantic_segmentation/DUC/README.md]: same as above

File: models/semantic_segmentation/DUC/README.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+Please note that the dataset is under copyright. Refer to the [citation](https://www.cityscapes-dataset.com/citation/) page for details.
+
+## Validation accuracy
+The [mIOU](#metric) score obtained by the models on the validation set are mentioned above and they match with those mentioned in the paper.
+
+## Training
+Coming soon.
+
+## <a name="metric"></a>Validation
+The metric used for validation is **mean Intersection Over Union (mIOU)**. For each class the intersection over union (IOU) of pixel labels between the output and the target segmentation maps is computed and then averaged over all classes to give us the mean intersection over union (mIOU).
```
Comment (aaronmarkham) [inline models/semantic_segmentation/DUC/README.md]: You should define this earlier. I read mIOU in the overview and didn't know what it was.

File: models/semantic_segmentation/DUC/README.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
+# Dense Upsampling Convolution (DUC)
+
+## Use cases
+DUC is a semantic segmentation model, i.e., for an input image the model labels each pixel in the image into a set of pre-defined categories. The model provides very good accuracy in terms of [mIOU](#metric) score and can be used in any application requiring semantic segmentation. In particular, since the model is trained on the [cityscapes dataset](#dset) which contains images from urban street scenes, it can be used effectively in self driving vehicle systems.
```
Comment (aaronmarkham) [inline models/semantic_segmentation/DUC/README.md]: please define mIOU here.

File: models/semantic_segmentation/DUC/cityscapes_labels.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/python
+#
+# Cityscapes labels
```
Comment (aaronmarkham) [inline models/semantic_segmentation/DUC/cityscapes_labels.py]: Usage info? What about adding some usage info to the README?

File: models/semantic_segmentation/DUC/cityscapes_loader.py (hunk 4 above)
Comment (aaronmarkham) [reply models/semantic_segmentation/DUC/cityscapes_loader.py]: Please include some overall usage information back in the README on how and when you use this script.

File: models/semantic_segmentation/DUC/README.md (hunk 5 above)
Comment (abhinavs95) [reply models/semantic_segmentation/DUC/README.md]: By earlier do you mean earlier in the readme for DUC? In order to stick to the readme template I defined the metric below and linked all previous instances of mIOU to that section.. does that work?

File: models/semantic_segmentation/DUC/cityscapes_loader.py (hunk 4 above)
Comment (abhinavs95) [reply models/semantic_segmentation/DUC/cityscapes_loader.py]: Added info about scripts in inference and validation sections of readme. Also added comments in the beginning of the scripts on where and what they are used for

File: models/semantic_segmentation/DUC/README.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+We used MXNet as framework to perform inference. View the notebook [duc-inference](duc-inference.ipynb) to understand how to use above models for doing inference. A brief description of the inference process is provided below:
+
+### Input 
+Since the model is trained on the cityscapes dataset which contains images of urban street scenes, the input should preferably be an image of a street scene to produce best results. There are no constraints on the size of the image. The example in the inference notebook is done using a png image.
+
+### Preprocessing
+The DUC layer has an effect of partitioning the image into d^2 subparts (d = downsampling rate). This is why the input image is extrapolated with a small border in order to obtain an accurate reshaped image after the DUC layer. After this the image is normalized using mean subtraction. Check [duc-preprocess.py](duc-preprocess.py) for code.
+
+### Output
+The output of the network is a tensor of shape (1 X `label_num` X `H`*`W`) where `H` and `W` are the height and width of the output segmented map.
```
[the remark this answers is not in the document]
Comment (abhinavs95) [reply models/semantic_segmentation/DUC/README.md]: done

File: models/semantic_segmentation/DUC/README.md (hunk 6 above)
Comment (abhinavs95) [reply models/semantic_segmentation/DUC/README.md]: done.. same as above

File: models/semantic_segmentation/DUC/cityscapes_labels.py (hunk 7 above)
Comment (abhinavs95) [reply models/semantic_segmentation/DUC/cityscapes_labels.py]: done!

**Comment by abhinavs95** [comment]
@aaronmarkham thanks for the review, made some changes addressing your comments

**Comment by ankkhedia** [comment]
Looks good to me.

**Comment by lupesko** [comment]
Looks good, thanks for this awesome contribution @abhinavs95 !
@ankkhedia can you please run the notebooks to make sure there is no issue? if all good, we are good to merge. Thanks!

**Comment by ankkhedia** [comment]
Verified the notebooks and merged.

**Pull request merged**